add a "memory management" documentation page (#11415)
apollographql/apollo-client: docs/source/caching/memory-management.mdx

---

> title: Memory management
> api_doc:
>   - "@apollo/client!CacheSizes:interface"
>   - "@apollo/client!ApolloClient:class"

Reviewer: ```suggestion
title: Memory management
api_doc:
  - "@apollo/client!CacheSizes:interface"
  - "@apollo/client!ApolloClient:class"
minVersion: 3.9.0
```

I'd add the `minVersion` here as well so that it shows up in the UI

You might also consider adding a blockquote in the intro mentioning its available in 3.9.0-beta.0?

> ---

Reviewer 2: ```suggestion
subtitle: Learn how to choose and set custom cache sizes
description: Learn how to choose and set custom cache sizes with Apollo Client.
---
```
(Latter lengthier `description` is for the meta-description/SEO purposes.)

import { Remarks, PropertySignatureTable, Example } from '../../shared/ApiDoc';

## Cache Sizes

For better performance, Apollo Client caches (or, in other words, memoizes) a lot
of internally calculated values.
In most cases, these values are cached in WeakCaches, which means that if the
source object is garbage-collected, the cached value will be garbage-collected,
too.

Additionally, those caches are LRU caches, which means that if the cache is full,
the least recently used value will be garbage-collected.

Depending on your application you might want to tweak the cache size to fit your
needs.

You can do this in two ways:

### Setting Cache Sizes before loading the Apollo Client library

This is the recommended way to set the cache sizes, as it will allow you to set
cache sizes before the Apollo Client library is loaded (some caches will already
be initialized when the library is loaded, and changed cache sizes will only

> affect caches created after the fact).

Author: Before: caches are created with the correct size.
After: caches are created with the wrong size, you change the size, you have to write code to recreate the cache.

=> More code and you potentially throw away work that has already been done before.

> This is the recommended way to set the cache sizes, as it will allow you to set
> cache sizes before the Apollo Client library is loaded (some caches will already
> be initialized when the library is loaded, and changed cache sizes will only
> affect caches created after the fact).

Author: ```suggestion
Setting cache sizes before loading the Apollo Client library is recommended because some caches are already initialized when the library is loaded, and changed cache sizes only
affect caches created after the fact, so you'd have to write additional runtime codes to recreate these caches after changing their size.
```

>  ```ts

Author: Not really, depending on the framework it could be an `main.ts`, `index.ts`, `entrypoint.ts` - really no common pattern there.

import type { CacheSizes } from '@apollo/client/utilities';

 globalThis[Symbol.for("apollo.cacheSize")] = {
  parser: 100,
  "fragmentRegistry.lookup": 500
 } satisfies Partial<CacheSizes>
 ```

### Adjusting Cache Sizes after loading the Apollo Client library

You can also adjust cache sizes after loading the library.

> ```js

Reviewer 2: Same as previous comment: does it make sense to add a file name?

Author: It can really be in any file the user might see fitting for it.

import { cacheSizes } from '@apollo/client/utilities';
import { print } from '@apollo/client'

cacheSizes.print = 100;

> // cache sizes changed this way will only take effect for caches created after

Author: I also added it to your suggestion earlier, but I believe it might be good to have this in two places - some people directly skip to the code examples, and this is very important information they must not miss.

// the cache size has been changed, so we need to reset the cache for it to be effective

print.reset();
```

> ### Choosing good cache sizes

Reviewer 2: ```suggestion
### Choosing appropriate cache sizes
```

<Remarks
  canonicalReference="@apollo/client!CacheSizes:interface"
/>

> To choose good sizes for our memoization caches, you need to know what they
> use as source values, and have a general understanding of the data flow inside of
> Apollo Client.

Reviewer 2: ```suggestion
To choose appropriate sizes for memoization caches, you need to know what the caches use as source values and understand data flow inside Apollo Client at a high level.
```

Can we add a relevant link for "understand data flow inside Apollo Client" to give folks a sense of what they need to know?

Author: That really should be an intro to the next paragraph that tries to give that overview :/ Do you have any idea how to bring that out better?

Reviewer 2: Ah. Well the next sentences describe the source types, but the "data flow" part is unclear for me. Is that what's described when talking about transforming inputs to outputs?

> 

Reviewer 2: ```suggestion

#### Cache source values

```

Author: Not sure if a heading makes sense here - the last paragraph directly feeds into this - this is said "overview" over the data flow.

> For most memoized values, the source value will be a parsed GraphQL document -
> a `DocumentNode`. Here, we need to distinguish between two types of documents:

Reviewer 2: ```suggestion
For most memoized values, the source value is a parsed GraphQL document&mdash;
a `DocumentNode`. There are two types:
```

> * User-supplied `DocumentNode`s: These are DocumentNode objects that are created

Reviewer 2: ```suggestion
* **User-supplied `DocumentNode`s** are created
```

  by the user, for example by using the `gql` template literal tag.

>   These are the `QUERY`, `MUTATION` or `SUBSCRIPTION` variables that you pass e.g.
>   into your `useQuery` hook, or as the `query` option to `client.query`.

Reviewer 2: ```suggestion
  These are the `QUERY`, `MUTATION`, or `SUBSCRIPTION` variables passed
  into a [`useQuery` hook](../data/queries/#usequery-api) or as the `query` option to `client.query`.
```

> * Transformed `DocumentNode`s: These are DocumentNode objects are derived from
>   user-supplied `DocumentNode`s, e.g. by applying `DocumentTransform`s to them.

Reviewer 2: ```suggestion
* *Transformed `DocumentNode`s** are derived from
  user-supplied `DocumentNode`s, for example, by applying [`DocumentTransform`s](../data/document-transforms/) to them.
```

As a rule of thumb, you should set the cache sizes for caches using a Transformed
`DocumentNode` at least to the same size as for caches using a user-supplied
`DocumentNode`. If your application uses a custom `DocumentTransform` that does
not always transform the same input to the same output, you should set the cache
size for caches using a Transformed `DocumentNode` to a higher value than for
caches using a user-supplied `DocumentNode`.

> By default, Apollo Client uses a "base value" of 1000 for caches using

Reviewer 2: Is "base value" the same as a default value? Is there a unit we can provide?

Author: The unit is "whatever we put into the cache" - we maybe could say "objects" of "cached values".
"Base value" here is the value we use to scale all other values of:
1000 for caches using user-provided DocumentNodes -> 2000 or 4000 for transformed ones.
For a base value of 500 it would be 
500 for caches using user-provided DocumentNodes -> 1000 or 2000 for transformed ones.

Reviewer 2: Got it, will put in a suggestion that reflects this info.

user-supplied `DocumentNode` instances, and scales other cache sizes relative
to that.

> This should be plenty for almost all applications out there, but you might want
> to tweak them if you have different requirements.

Reviewer 2: ```suggestion
This base value should be plenty for most applications, but you can tweak them if you have different requirements.
```

Beyond this copyedit, I think it makes sense to start this section with information about the base/default and then explain why users may need to change the default. (What does "have different requirements" mean?)

#### Measuring cache usage

> As it can be hard to estimate good cache sizes for your application, Apollo Client

Reviewer 2: ```suggestion
Since estimating appropriate cache sizes for your application can be hard, Apollo Client
```

exposes an API for cache usage measurement.<br />
This way, you can click around in your application and then take a look at the
actual usage of the memoizing caches.

> Keep in mind that this API is primarily meant for usage with our DevTools

Reviewer 2: ```suggestion
Keep in mind that this API is primarily meant for usage with the Apollo DevTools
```

> (we will release an integration for this soon), and we might change it at any

Reviewer 2: ```suggestion
(an integration is coming soon), and the API may change at any
```

point in time.<br />
It is also only included in development builds, not in production builds.

> So please only use this for manual measurements, and don't rely on it in production
> code or tests.
> 

Reviewer 2: ```suggestion
<Note>

The cache usage API is only meant for manual measurements. Don't rely on it in production code or tests.

</Note>

```

<Example
  canonicalReference="@apollo/client!ApolloClient#getMemoryInternals:member"
  index={0}
/>

<Example
  collapsible
  canonicalReference="@apollo/client!ApolloClient#getMemoryInternals:member"
  index={1}
/>

### Cache options

> <PropertySignatureTable

Reviewer 2: Is it intended that the first few options are in `" "`?

Author: Yes, they contain a dot, so to be used as valid JavaScript properties, they have to be used as a string.

---
title: Memory management
api_doc:
  - "@apollo/client!CacheSizes:interface"
  - "@apollo/client!ApolloClient:class"
subtitle: Learn how to choose and set custom cache sizes
description: Learn how to choose and set custom cache sizes with Apollo Client.
---

import { Remarks, PropertySignatureTable, Example } from '../../shared/ApiDoc';

## Cache Sizes

For better performance, Apollo Client caches (or, in other words, memoizes) many
internally calculated values.
In most cases, these values are cached in [weak caches](https://en.wikipedia.org/wiki/Weak_reference), which means that if the
source object is garbage-collected, the cached value will be garbage-collected,
too.

These caches are also Least Recently Used (LRU) caches, meaning that if the cache is full,
the least recently used value will be garbage-collected.

Depending on your application, you might want to tweak the cache size to fit your
needs.

> You can set your cache size [before (recommended)](#set-cache-sizes-before-loading-the-apollo-client-library) or [after](#adjust-cache-sizes-after-loading-the-apollo-client-library) loading the Apollo Client library.

Reviewer: ```suggestion
You can set your cache size [before (recommended)](#setting-cache-sizes-before-loading-the-apollo-client-library) or [after](#adjusting-cache-sizes-after-loading-the-apollo-client-library) loading the Apollo Client library.
```

### Setting cache sizes before loading the Apollo Client library

> Setting cache sizes before loading the Apollo Client library is recommended because some caches are already initialized when the library is loaded, and changed cache sizes only
> affect caches created after the fact, so you'd have to write additional runtime codes to recreate these caches after changing their size.

Reviewer: ```suggestion
Setting cache sizes before loading the Apollo Client library is recommended because some caches are already initialized when the library is loaded. Changed cache sizes only
affect caches created after the fact, so you'd have to write additional runtime codes to recreate these caches after changing their size.
```

I think splitting this into two sentences might feel a bit nicer so it doesn't end up as a run on sentence.

 ```ts
import type { CacheSizes } from '@apollo/client/utilities';

 globalThis[Symbol.for("apollo.cacheSize")] = {
  parser: 100,
  "fragmentRegistry.lookup": 500
 } satisfies Partial<CacheSizes>
 ```

### Adjusting cache sizes after loading the Apollo Client library

You can also adjust cache sizes after loading the library.

```js
import { cacheSizes } from '@apollo/client/utilities';
import { print } from '@apollo/client'

cacheSizes.print = 100;

> // cache sizes changed this way will only take effect for caches created after
> // the cache size has been changed, so we need to reset the cache for it to be effective

Reviewer: ```suggestion
// cache sizes changed this way will only take effect for caches 
// created after the cache size has been changed, so we need to 
// reset the cache for it to be effective.
```

To avoid horizontal scroll, I'd break this up just a bit more

print.reset();
```

### Choosing appropriate cache sizes

<Remarks
  canonicalReference="@apollo/client!CacheSizes:interface"
/>

To choose good sizes for our memoization caches, you need to know what they
use as source values, and have a general understanding of the data flow inside of
Apollo Client.

For most memoized values, the source value is a parsed GraphQL document&mdash;
a `DocumentNode`. There are two types:

* **User-supplied `DocumentNode`s** are created
  by the user, for example by using the `gql` template literal tag.

>   These are the `QUERY`, `MUTATION`, or `SUBSCRIPTION` variables passed

Reviewer: ```suggestion
  This is the `QUERY`, `MUTATION`, or `SUBSCRIPTION` argument passed
```

I think "variables" might confused people here who think this means the `variables` option.

  into a [`useQuery` hook](../data/queries/#usequery-api) or as the `query` option to `client.query`.

> * *Transformed `DocumentNode`s** are derived from

Reviewer: ```suggestion
* **Transformed `DocumentNode`s** are derived from
```

/>

To choose good sizes for our memoization caches, you need to know what they
use as source values, and have a general understanding of the data flow inside of
Apollo Client.

For most memoized values, the source value is a parsed GraphQL document&mdash;
a `DocumentNode`. There are two types:

* **User-supplied `DocumentNode`s** are created
  by the user, for example by using the `gql` template literal tag.
  This is the `QUERY`, `MUTATION`, or `SUBSCRIPTION` argument passed
  into a [`useQuery` hook](../data/queries/#usequery-api) or as the `query` option to `client.query`.
* **Transformed `DocumentNode`s** are derived from
  user-supplied `DocumentNode`s, for example, by applying [`DocumentTransform`s](../data/document-transforms/) to them.

As a rule of thumb, you should set the cache sizes for caches using a transformed
`DocumentNode` at least to the same size as for caches using a user-supplied
`DocumentNode`. If your application uses a custom `DocumentTransform` that does
not always transform the same input to the same output, you should set the cache
size for caches using a Transformed `DocumentNode` to a higher value than for
caches using a user-supplied `DocumentNode`.

> By default, Apollo Client uses a "base value" of 1000 for caches using
> user-supplied `DocumentNode` instances, and scales other cache sizes relative
> to that.

Reviewer 2: Hopefully, I'm understanding correctly and this is an accurate description:

```suggestion
By default, Apollo Client uses a base value of 1000 cached objects for caches using
user-supplied `DocumentNode` instances, and scales other cache sizes relative
to that. For example, the default base value of 1000 for user-provided `DocumentNode`s would scale to 2000, 4000, etc. for transformed `DocumentNode`s, depending on the transformation performed.
```

Reviewer: Looks good to me!
